File: online-game-store-front/src/app/classes/game.ts

import { Guid } from "guid-typescript";
import * as moment from "moment";
import { Moment } from "moment";
import { Genre } from "./genre";
import { PlatformType } from "./platformType";
import { Publisher } from "./publisher";

export class Game {
  [prop: string]: any;
  constructor(
  public id: string = Guid.EMPTY,
  public key: string = '',
  public name: string = '',
  public description: string = '',
  public viewsNumber: number = 0,
  public commentsNumber: number = 0,
  public added: Moment = moment.utc(),
  public published: Moment = moment.utc(),
  public price: number = 0,
  public discount: number = 0,
  public unitsInStock: number = 0,
  public discontinued: boolean = false,
  public isCommented: boolean = true,
  public isDownloaded: boolean = true,
  public publisher: Publisher | null = null,
  public genres: Genre[] = [],
  public platformTypes: PlatformType[] = []
  ) { }
}

File: online-game-store-api/OnlineGameStore.DAL/Entities/Game.cs
using MongoDB.Bson.Serialization.Attributes;
using OnlineGameStore.DAL.BsonSerializers;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.RegularExpressions;

namespace OnlineGameStore.DAL.Entities
{
    public class Game : BaseEntity
    {
        [BsonElement("ProductID")]
        [BsonSerializer(typeof(IntToGuidBsonSerializer))]
        public override Guid Id { get; set; }

        private string key;
        public string Key
        {
            get => key;
            set => key = Regex.Replace(value ?? Name, @"[^0-9a-zA-Z]+", "-").ToLower();
        }

        [BsonElement("CategoryID")]
        [NotMapped]
        [BsonSerializer(typeof(IntToNullableGuidBsonSerializer))]
        public Guid? MongoGenreId { get; set; }


        [BsonElement("ProductName")]
        public string Name { get; set; }

        [BsonElement("QuantityPerUnit")]
        public string Description { get; set; }
        public int ViewsNumber { get; set; }
        public DateTime Published { get; set; }
        public int UnitsOnOrder { get; set; }
        public int ReorderLevel { get; set; }
        public string DownloadPath { get; set; }

        private decimal price;
        [BsonElement("UnitPrice")]
        public decimal Price
        {
            get => Math.Round(price, 2);
            set
            {
                if (value < 0)
                {
                    throw new ArgumentException("Price must be grater than 0");
                }
                else
                {
                    price = Math.Round(value, 2);
                }
            }
        }

        private short unitsInStock;
        public short UnitsInStock
        {
            get => unitsInStock;
            set
            {
                if (value < 0)
                {
                    throw new ArgumentException("Units in stock must be grater than 0");
                }
                else
                {
                    unitsInStock = value;
                }
            }
        }

        private short discount;
        public short Discount
        {
            get => discount;
            set
            {
                if (value < 0 && value > 100)
                {
                    throw new ArgumentException("Discount must be in 0 - 100 range");
                }
                else
                {
                    discount = value;
                }
            }
        }

        public bool Discontinued { get; set; }


        [BsonElement("SupplierID")]
        [BsonSerializer(typeof(IntToNullableGuidBsonSerializer))]
        public Guid? PublisherId { get; set; }
        public virtual Publisher Publisher { get; set; }

        public virtual ICollection<GameGenre> Genres { get; set; }
        public virtual ICollection<Comment> Comments { get; set; }
        public virtual ICollection<GamePlatformType> PlatformTypes { get; set; }
    }
}